src/bson.rs

```diff
@@ -304,59 +304,108 @@ impl From<Value> for Bson {
         match a {
             Value::Number(x) => x
                 .as_i64()
-                .map(Bson::from)
+                .map(|i| {
+                    if i >= i32::MIN as i64 && i <= i32::MAX as i64 {
```

Comment:
Previously, we were reading converting every JSON integer to an i64. The spec requires that we use the smallest size integer possible, so this logic checks to see if it can fit in an i32.